src/coreclr/vm/jitinterface.cpp

```diff
@@ -7239,28 +7239,92 @@ bool getILIntrinsicImplementationForInterlocked(MethodDesc * ftn,
     if (ftn->GetMemberDef() != CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_T)->GetMemberDef())
         return false;
 
-    // Get MethodDesc for non-generic System.Threading.Interlocked.CompareExchange()
-    MethodDesc* cmpxchgObject = CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_OBJECT);
-
-    // Setup up the body of the method
-    static BYTE il[] = {
-                          CEE_LDARG_0,
-                          CEE_LDARG_1,
-                          CEE_LDARG_2,
-                          CEE_CALL,0,0,0,0,
-                          CEE_RET
-                        };
-
-    // Get the token for non-generic System.Threading.Interlocked.CompareExchange(), and patch [target]
-    mdMethodDef cmpxchgObjectToken = cmpxchgObject->GetMemberDef();
-    il[4] = (BYTE)((int)cmpxchgObjectToken >> 0);
-    il[5] = (BYTE)((int)cmpxchgObjectToken >> 8);
-    il[6] = (BYTE)((int)cmpxchgObjectToken >> 16);
-    il[7] = (BYTE)((int)cmpxchgObjectToken >> 24);
+    // Determine the type of the generic T method parameter
+    _ASSERTE(ftn->HasMethodInstantiation());
+    _ASSERTE(ftn->GetNumGenericMethodArgs() == 1);
+    TypeHandle typeHandle = ftn->GetMethodInstantiation()[0];
+
+    // Based on the generic method parameter, determine which overload of CompareExchange
+    // to delegate to, or if we can't handle the type at all.
+    int ilIndex = 0;
+    if (typeHandle.IsValueType())
+    {
+        if (!CorTypeInfo::IsPrimitiveType(typeHandle.GetVerifierCorElementType()))
+            return false;
+
+        switch (typeHandle.GetSize())
+        {
+            case 1: ilIndex = 1; break;
+            case 2: ilIndex = 1; break;
+            case 4: ilIndex = 1; break;
+            case 8: ilIndex = 1; break;
+            default: _ASSERT(!"Unexpected primitive type size"); break;
+        }
+    }
+
+    // Setup up the body of the CompareExchange methods
+    mdMethodDef cmpxchgTokenObject = CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_OBJECT)->GetMemberDef();
```

Comment:
It would be better to move the statics into to switch statement so that we do not have to fetch 5 methods each time even though we really only need one.

Reply:
Ok